### PR TITLE
Update activation GC page

### PR DESCRIPTION
**Pull request by sergeybykov**
Updated activation GC page to document and emphasize programmatic configuration and, hopefully, for better clarity.

### CODE REVIEW

**Comment by yevhen** [comment]
👍

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 1]
```diff
@@ -36,12 +27,77 @@ The most common approach is to configure the silo to periodically scan for activ
...
+
+### Delaying Activation GC
+
+A grain activation can delay its own Activation GC, by calling `this.DelayDeactivation()` method:
+
+``` csharp
+protected void DelayDeactivation(TimeSpan timeSpan)
+```
+
+This call will insure that this activations is not deactivated for at least the specified time duration. It takes priority over Activation Garbage Collection settings specified in the config, but does not cancel them.
```
Comment (ReubenBond) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: nit: ensure

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 2]
```diff
@@ -36,12 +27,77 @@ The most common approach is to configure the silo to periodically scan for activ
...
+Therefore, this call provides an additional hook to **delay the deactivation beyond what is specified in the Activation Garbage Collection settings**. This call can not be used to expedite Activation Garbage Collection.
+
+
+A positive <c>`timeSpan`</c> value means “prevent GC of this activation for that time span”.
+
+A negative <c>`timeSpan`</c> value means “cancel the previous setting of the `DelayDeactivation` call and make this activation behave based on the regular Activation Garbage Collection settings ”.
```
Comment (ReubenBond) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: nit: superfluous space before final "

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 3]
```diff
@@ -36,12 +27,77 @@ The most common approach is to configure the silo to periodically scan for activ
...
+
+2) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(5 min), the activation will be collected after 10 min, if no extra calls were made.
+
+3) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(5 min), and after 7 minutes there is another call on this grain, the activation will be collected after 17 min from time zero, if no extra calls were made.
+
+4) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(20 min), and after 7 minutes there is another call on this grain, the activation will be collected after 20 min from time zero, if no extra calls were made.
+
+Note that `DelayDeactivation` does not 100% guarantee that the grain activation will not get activated before the specified period of time expires. There are certain failure cases that may cause 'premature' deactivation of grains.
```
Comment (ReubenBond) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: nit: not get deactivated

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 4]
```diff
@@ -4,27 +4,18 @@ title: Activation Garbage Collection
...
+Activation Garbage Collection (Activation GC) is the process of removal from memory of unused grain activations. It is conceptually similar to how garbage collection of memory works in .NET. However, Activation GC only takes into consideration how long a particular grain activation has been idle. Memory usage is not used as a factor.
 
- In the past, we experimented with Activation-GC based on memory usage triggers, but we abandoned that approach due to the unreliable memory usage metrics we could use.
+## How Activation GC Works
+The general process of Activation GC involves Orleans runtime in a silo periodically scanning for grain activations that have not been used at all for the configured period of time (Collection Age Limit). Once a grain activation has been idle for that long, it gets deactivated. The deactivation process begins by the runtime calling the grain’s `OnDeactivate()` method, and completes by removing references to the grain activation object from all data structures of the silo, so that the memory is reclaimed by the .NET GC.
```
Comment (jdom) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: I believe it's `OnDeactivateAsync()`, isn't it?

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 5]
```diff
@@ -36,12 +27,73 @@ The most common approach is to configure the silo to periodically scan for activ
...
+Therefore, this call provides an additional hook to **delay the deactivation beyond what is specified in the Activation Garbage Collection settings**. This call can not be used to expedite Activation Garbage Collection.
+
+
+A positive <c>`timeSpan`</c> value means “prevent GC of this activation for that time span”.
+
+A negative <c>`timeSpan`</c> value means “cancel the previous setting of the `DelayDeactivation` call and make this activation behave based on the regular Activation Garbage Collection settings”.
+
+**Scenarios:**
+
+1) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(20 min), it will cause this activatin to not be collected for at least 20 min.
```
Comment (jdom) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: typo: activatin

File: src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md [hunk 6]
```diff
@@ -36,12 +27,73 @@ The most common approach is to configure the silo to periodically scan for activ
...
+
+2) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(5 min), the activation will be collected after 10 min, if no extra calls were made.
+
+3) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(5 min), and after 7 minutes there is another call on this grain, the activation will be collected after 17 min from time zero, if no extra calls were made.
+
+4) Activation Garbage Collection settings specify age limit of 10 minutes and the grain is making a call to DelayDeactivation(20 min), and after 7 minutes there is another call on this grain, the activation will be collected after 20 min from time zero, if no extra calls were made.
 
- The runtime's default collection age limit is 2 hours.
+Note that `DelayDeactivation` does not 100% guarantee that the grain activation will not get deactivated before the specified period of time expires. There are certain failure cases that may cause 'premature' deactivation of grains.
```
Comment (jdom) [inline src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: this note is something VERY important that people sometime miss, and think they can pin a grain activation to a silo forever. Maybe you want to be a little bit more explicit about what are the failure scenarios, or at least highlight that delaying deactivation should not be used to make an activation exist forever during the lifetime of the silo it lives in.

Comment (sergeybykov) [reply src/Documentation/Advanced-Concepts/Activation-Garbage-Collection.md]: Good idea. I added a couple of sentences to stress that.

**Pull request merged**